sklearn/tree/tests/test_tree.py

```diff
@@ -831,7 +841,58 @@ def test_min_impurity_split():
                                       est.min_impurity_split))
 
 
-def test_pickle():
+def test_min_impurity_decrease():
+    # test if min_impurity_decrease ensure that a split is made only if
+    # if the impurity decrease is atleast that value
+    X = np.asfortranarray(iris.data.astype(tree._tree.DTYPE))
+    y = iris.target
+
+    # test both DepthFirstTreeBuilder and BestFirstTreeBuilder
+    # by setting max_leaf_nodes
+    for max_leaf_nodes, name in product((None, 1000), ALL_TREES.keys()):
+        TreeEstimator = ALL_TREES[name]
+
+        # Check default value of min_impurity_decrease, 1e-7
+        est1 = TreeEstimator(max_leaf_nodes=max_leaf_nodes,
+                            random_state=0)
+        # Check with explicit value of 0.05
+        est2 = TreeEstimator(max_leaf_nodes=max_leaf_nodes,
+                            min_impurity_decrease=0.05,
+                            random_state=0)
+        # Check with a much lower value of 0.00001
+        est3 = TreeEstimator(max_leaf_nodes=max_leaf_nodes,
+                            min_impurity_decrease=0.00001,
+                            random_state=0)
+
+        for est, expected_decrease in ((est1, 1e-7), (est2, 0.05),
+                                       (est3, 0.00001)):
+            assert_less_equal(est.min_impurity_decrease, expected_decrease,
+                              "Failed, min_impurity_decrease = {0} > {1}"
+                              .format(est.min_impurity_decrease,
+                                      expected_decrease))
+            est.fit(X, y)
+            for node in range(est.tree_.node_count):
+                # If current node is a not leaf node, check if the split was
+                # justified w.r.t the min_impurity_decrease
+                if est.tree_.children_left[node] != TREE_LEAF:
+                    left = est.tree_.children_left[node]
+                    imp_left = est.tree_.impurity[left]
+                    weighted_n_left = est.tree_.weighted_n_node_samples[left]
+
+                    right = est.tree_.children_right[node]
+                    imp_right = est.tree_.impurity[right]
+                    weighted_n_right = est.tree_.weighted_n_node_samples[right]
+
+                    actual_decrease = (est.tree_.impurity[node] -
```

Comment:
#TODO this is incorrect comparison. The actual decrease should again by multiplied by fractional weight of the parent node...